### PR TITLE
Test mixer

### PR DESCRIPTION
**Pull request by kyessenov**
**What this PR does / why we need it**:
Introduces a test mixer that echoes received attributes to stdout. The plan is to mock mixer behavior and add pre-defined check policies, for testing pilot/proxy behavior.
Lightens the load on the test cluster as the test mixer is very lightweight at the moment.
The tests can now validate that all attributes are set correctly, without requiring full-fledged mixer processing.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: 
contributed to #533

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by kyessenov** [comment]
I wanted to avoid a direct dependency on mixer codebase since it's moving fast, and I'm not using most of Mixer functionality in my test. I think the ideal solution is to pimp up istio/api with codegen stuff. Is that what you propose? Note that most of sed-ding is to get rid of gogoproto annotations since they're very hard to build with current rules_go.

**Comment by costinm** [comment]
Don't we have a test mixer in the proxy integration tests ? If they serve
different purpose it's ok to have 2, but
should probably be shared, it doesn't make sense to have one test style per
git repo.

istio/istio has common test infra - may be the best place.

**Comment by kyessenov** [comment]
They serve different purpose.

**Comment by douglas-reid** [comment]
improving istio/api sounds good. i was also hopeful that we could have a forcing function to improve `@io_istio_mixer//test` such that users wouldn't be impacted by changes to mixer proper.  but we can work on that separately.

File: test/mixer/server.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	bag := s.decode(msg.Attributes)
+	fmt.Printf("[%d] words.size=%d\n", s.counter, len(msg.Attributes.Words))
+	for k, v := range bag {
+		fmt.Printf("[%d] %s=%s\n", s.counter, k, v)
+	}
+	return &istio_mixer_v1.CheckResponse{
+		Precondition: &istio_mixer_v1.CheckResponse_PreconditionResult{
+			Status: &google_rpc.Status{Code: 0},
+		},
+	}, nil
```
Comment (rshriram) [inline test/mixer/server.go]: Is it possible to mock this or add a parameter to this such that we can decide when Mixer should fail or accept a check request? Would come in handy during testing stability

File: test/mixer/wordlist_process.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+cat <<EOF
```
Comment (rshriram) [inline test/mixer/wordlist_process.sh]: missing copyright..
Also, please add a comment on what this shell script does.

File: test/mixer/pb/proto_process.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
Comment (rshriram) [inline test/mixer/pb/proto_process.sh]: copyright

**Comment by rshriram** [comment]
Largely LGTM, missing copyrights in couple of files. And some high level comment on the shell script that is generating go code. I thought you hated shell scripts!
Might also want to update our docs (testing.md) stating that we use a mock mixer for tests.

**Comment by rshriram** [comment]
and I am going by @douglas-reid 's reviews for the mixer replacement. So, if the replacement is good with mixer folks, fine by me.

File: test/mixer/server.go (hunk 1 above)
Comment (kyessenov) [reply test/mixer/server.go]: Yeah, we can do absolutely anything now that we can have a server endpoint. This is just a start to get data path validation, before enforcing control behavior. I'm thinking of adding control HTTP endpoints on this to mark start/end of a test, and then send rules for asserting check response.

File: test/mixer/wordlist_process.sh (hunk 2 above)
Comment (kyessenov) [reply test/mixer/wordlist_process.sh]: This is auto-generated code. I'll add one copyright on the script.

File: test/mixer/pb/proto_process.sh (hunk 3 above)
Comment (kyessenov) [reply test/mixer/pb/proto_process.sh]: Same

**Comment by costinm** [comment]
So is the expectation that any project using mixer will create its own mock
mixer ? Or will they depend on pilot (or proxy) and reuse their mixer ?

Depending on mixer repo will be required for any project developing plugins
- given the WORKSPACE diffs, it would be quite tricky to depend on
both mixer and pick the mock from pilot repo.

istio/api or istio/istio seem better places for code test infra that may be
reused - the fact proxy also needs a mock mixer suggest it's going to be
a common need. I suppose a mock pilot will be great to have in the same
place as well.

**Comment by kyessenov** [comment]
The more mixers in the world the better. The protocol is easy to implement, and we need wire-level observation to validate that we program mixerclient correctly. 

Mixer plugins have entirely different needs. They will use templates and bazel machinery, and should be disallowed access to raw attributes.

**Comment by kyessenov** [comment]
PTAL. Added copyrights and documentation for shell scripts.

**Comment by rshriram** [comment]
@ZackButcher this is non functional change, to improve our test suites for better mixer tests. @douglas-reid LGTM'ed it as well.

**Pull request merged**